### DISCUSSION TITLE
Should assignment to a observable value become an observable?

### DISCUSSION THREAD

**Discussion by wegry** [mobxjs/mobx#2879] [Q&A] [answered]
Hey, we're currently using mobx versions like this https://codesandbox.io/s/zealous-volhard-20zt1?file=/src/App.tsx.

Should assigning regular object to a proxied observable ~create~ make the _newly assigned_ object observable?

It seems like in this example, the assigned value has lost its observability (observerness)?

**Comment by urugator** [comment, marked as answer]
> Should assigning regular object to a proxied observable create make the object observable?

If you assign *non-observable plain object* to observable object, it's automatically copied and converted to observable. 
`store.foo = { x: 1 }` is same as `store.foo = observable({ x: 1 })` 
The behavior can be modified with `observable.ref` annotation or `deep:  false` option.
Not sure if that's what you're asking about.

**Reply by wegry** [reply]
@urugator I think you answered my question.

My code sandbox shows that isn't happening though, I think. The console.debugs after updates aren't proxied objects. Unless proxied objects `console.debug` without the proxy at the top level.

```typescript
class Store {
  currentInstance: { foo: string } & Record<string, undefined | string> = {
    otherField: "so defined",
    foo: "so defined"
  };
  constructor() {
    makeObservable(this, { currentInstance: observable });
  }
}

export default observer(function App() {
  const localObservable = useLocalObservable(() => new Store());

  useEffect(() => {
    runInAction(() => {
      localObservable.currentInstance = {
        foo: new Date().toISOString()
      };
    });
    console.debug(
      "inside the interval",
      localObservable.currentInstance,
      localObservable.currentInstance.foo
    );
  }, []);

  console.debug(
    "outside the interval",
    localObservable.currentInstance,
    localObservable.currentInstance.foo
  );
```

logs

```
outside the interval 

    {otherField: "so defined", foo: "so defined"}

so defined 
inside the interval 

    {foo: "2021-04-08T16:09:32.534Z"}

2021-04-08T16:09:32.534Z 
outside the interval 

    {foo: "2021-04-08T16:09:32.534Z"}

2021-04-08T16:09:32.534Z 
```

Maybe I'm using the annotation incorrectly in the constructor though.

**Reply by mweststrate** [reply]
@wegry that output looks correct to me? Did you expect something else?

**Reply by wegry** [reply]
@mweststrate when we log `observable` fields locally, they're usually `Proxy`s _and not the wrapped object without using `toJS`._

<img width="409" alt="Screen Shot 2021-04-08 at 12 26 40 PM" src="https://user-images.githubusercontent.com/5156873/114062649-bfaf7d00-9865-11eb-888a-631d4c70e848.png">

Maybe we're misconfigured though. We don't support IE 11 or use decorators.

*edit:* tweaking the browserlist in the code sandbox's package.json to not support IE 11 now `console.debug`s the top level proxy like how our local setup works.

**Reply by mweststrate** [reply]
ok but that looks like a different tool? I think you are now debugging differences between console.debug implementations, not differences in MobX? Given that your component is re-rendering after the effect seems an indication that everything is ok. You can use reflection utilities like `isObservableObject` to be sure (whether an object is a Proxy or not is in principle unobservable from JS, so a tool like code sandbox is technically not even able to tell you whether it is a proxy or not)

**Reply by wegry** [reply]
> I think you are now debugging differences between console.debug implementations, not differences in MobX?

It was an errant browserslist setting in the package.json containing IE 11 support. My mistake. The proxy is logging as expected now.

@mweststrate thanks for pointing us toward `isObservableObject`, hadn't seen that yet. 

A little bit more context is we have a relatively deeply nested object coming back from https://github.com/rjsf-team/react-jsonschema-form.

```diff
diff --git a/file.tsx b/file.tsx
--- a/file.tsx
+++ b/file.tsx
   const onChange = useCallback(
     (val: any) => {
       tweakOnChange?.({ value: val, schema })
-      modelStore.currentModelInstance = val
+     Object.assign(modelStore.currentModelInstance, val)
     },
     [modelStore, schema, tweakOnChange]
   )
```

Without the Object.assign, assigning the variable `val` (which isn't observable) to `modelStore.currentModelInstance` (which should be observable), doesn't seem to trigger an update.

**Comment by mweststrate** [comment]
It sounds like your currentModelInstance field itself is not marked as
observable, or that reference is being read by a component that is not an
observer.
